### ISSUE TITLE
memory leak 

### ISSUE THREAD

**Issue by f4nff** [blechschmidt/tun2proxy#32]
```
==87237== 663,044,096 bytes in 1 blocks are possibly lost in loss record 31 of 31 
==87237==    at 0x483DFAF: realloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so) 
==87237==    by 0x184631: _ZN5alloc7raw_vec11finish_grow17h42006598bf6c1d6cE.llvm.17228478732513752358 (in /root/tun2proxy/target/release/tun2proxy) 
==87237==    by 0x135BA0: alloc::raw_vec::RawVec<T,A>::reserve::do_reserve_and_handle (in /root/tun2proxy/target/release/tun2proxy) 
==87237==    by 0x189738: <alloc::collections::vec_deque::VecDeque<T,A> as core::iter::traits::collect::Extend<&T>>::extend (in /root/tun2proxy/target/release/tun2proxy) 
==87237==    by 0x1798BD: tun2proxy::http::HttpConnection::state_change (in /root/tun2proxy/target/release/tun2proxy) 
==87237==    by 0x179E44: <tun2proxy::http::HttpConnection as tun2proxy::tun2proxy::TcpProxy>::push_data (in /root/tun2proxy/target/release/tun2proxy) 
==87237==    by 0x16BDB4: tun2proxy::tun2proxy::TunToProxy::run (in /root/tun2proxy/target/release/tun2proxy) 
==87237==    by 0x164909: tun2proxy::main_entry (in /root/tun2proxy/target/release/tun2proxy) 
==87237==    by 0x14E699: tun2proxy::main (in /root/tun2proxy/target/release/tun2proxy) 
==87237==    by 0x142C02: std::sys_common::backtrace::__rust_begin_short_backtrace (in /root/tun2proxy/target/release/tun2proxy) 
==87237==    by 0x142658: _ZN3std2rt10lang_start28_$u7b$$u7b$closure$u7d$$u7d$17hd6fdf1e226452e52E.llvm.18146766573043479346 (in /root/tun2proxy/target/release/tun2proxy) 
==87237==    by 0x2A9D5B: call_once<(), (dyn core::ops::function::Fn<(), Output=i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe)> (function.rs:287) 
==87237==    by 0x2A9D5B: do_call<&(dyn core::ops::function::Fn<(), Output=i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe), i32> (panicking.rs:487) 
==87237==    by 0x2A9D5B: try<i32, &(dyn core::ops::function::Fn<(), Output=i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe)> (panicking.rs:451) 
==87237==    by 0x2A9D5B: catch_unwind<&(dyn core::ops::function::Fn<(), Output=i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe), i32> (panic.rs:140) 
==87237==    by 0x2A9D5B: {closure#2} (rt.rs:148) 
==87237==    by 0x2A9D5B: do_call<std::rt::lang_start_internal::{closure_env#2}, isize> (panicking.rs:487) 
==87237==    by 0x2A9D5B: try<isize, std::rt::lang_start_internal::{closure_env#2}> (panicking.rs:451) 
==87237==    by 0x2A9D5B: catch_unwind<std::rt::lang_start_internal::{closure_env#2}, isize> (panic.rs:140) 
==87237==    by 0x2A9D5B: std::rt::lang_start_internal (rt.rs:148)
```

**Comment by blechschmidt** [comment]
ad72147ff40320af145550018a18d4000bc41e86 fixes two memory issues. Valgrind is still not to happy about some places, but I currently don't know why. I will keep observing this.

**Comment by f4nff** [comment]
![image](https://user-images.githubusercontent.com/19538878/235033782-68ad2f66-b88f-4c41-a989-8692bcd3f3f4.png)

There is still a memory leak,

**Comment by jorgeajimenezl** [comment]
One thing I noticed after reading the code is that on clients where the exchange between the proxy and the client is at different speeds, tun2proxy can allocate large amounts of memory.

**Comment by blechschmidt** [comment]
You are right. Especially in upload scenarios, large uploads will cause the proxy clients to consume all data from the smoltcp sockets on the tun interface and put it into the buffers for the mio sockets that communicate with the proxy. A solution to this problem would be to use a fixed-size buffer for the mio clients (that is as large as the smoltcp buffer) and only consume at most the number of bytes that can be written to the mio client buffers.

**Comment by ssrlive** [comment]
@f4nff please test this branch https://github.com/blechschmidt/tun2proxy/tree/memory 
if you looks fine, i will merge it to master branch.

**Comment by ssrlive** [comment]
@f4nff and @jorgeajimenezl   please test this branch https://github.com/blechschmidt/tun2proxy/tree/memory
if you looks fine, i will merge it to master branch.

**Comment by ssrlive** [comment]
fixed in #77
